include/cppflow/tensor.h

```diff
@@ -196,26 +207,31 @@ namespace cppflow {
 
     template<typename T>
     std::vector<T> tensor::get_data() const {
-        auto res_tensor = TFE_TensorHandleResolve(this->tfe_handle.get(), context::get_status());
-        status_check(context::get_status());
+        auto res_tensor = get_tensor();
 
         // Check tensor data is not empty
-        auto raw_data = TF_TensorData(res_tensor);
+        auto raw_data = TF_TensorData(res_tensor.get());
         //this->error_check(raw_data != nullptr, "Tensor data is empty");
 
-        size_t size = TF_TensorByteSize(res_tensor) / TF_DataTypeSize(TF_TensorType(res_tensor));
+        size_t size = TF_TensorByteSize(res_tensor.get()) / TF_DataTypeSize(TF_TensorType(res_tensor.get()));
 
         // Convert to correct type
         const auto T_data = static_cast<T*>(raw_data);
         std::vector<T> r(T_data, T_data + size);
-        TF_DeleteTensor(res_tensor);
 
         return r;
     }
 
     datatype tensor::dtype() const {
         return TFE_TensorHandleDataType(this->tfe_handle.get());
     }
+    
+    inline std::shared_ptr<TF_Tensor> tensor::get_tensor() const {
+        // NOTE: If the behavior of test1-3 in examples/tensor/main.cpp is reliable (unlikely at this time), we should be able to optimize the TFE_TensorHandleResolve call
+        tf_tensor = {TFE_TensorHandleResolve(tfe_handle.get(), context::get_status()), TF_DeleteTensor};
```

Comment:
I think `tf_tensor` should be a local variable but not a member. It will generate a new `TF_tensor` when `get_tensor` called every time. Actually, it's not a real cache.

Reply:
Yes and no...

No because it is not the data cache pairing the `tfe_handle` and getting updated on demand. The currently behavior simply forces people call `TFE_TensorHandleResolve`.

Yes, because test1-3 in examples/tensor/main.cpp. As I noted in the comments, it works if we only call `TFE_TensorHandleResolve` when `tf_tensor` is nullptr. I guess it requires the device is local. If one modifies `tf_tensor`, the changes are reflected in `tfe_handle`. However, we don't know how to synchronize this cache with `tfe_handle`. So it is a cache, but I don't want to rely on tensorflow internal implementation.

I am also fine If we remove `tf_tensor` from `tensor` class, but it creates lifetime management issues when calling some ops, e.g. some ops in raw_ops.h. The current `tf_tensor` is a cache for those ops too. Again, this is not ideal either, because calling `get_tensor()` again likely cause problems. I think the right way is to store the `tf_tensor` in the output `tensor` of those ops.

If you have a better implementation, feel free to make a PR :)

Reply:
@ivanallen I changed my mind... It seems relying on the internal TF cache is better. (code updated already) Once `tf_tensor` and `tfe_handle` are bound together, the `tf_tensor` cache does not change anymore. It refers to the local mirror in `tfe_handle`.

@serizba Can you also have a look at this?

Reply:
It makes sense and is good for performance!

I have a question here:

Why does not set get_tensor private? It's a little bit unreasonable that `struct TF_tensor` is visible for users.

Reply:
There are other classes and functions using `tf_tensor`, e.g. `raw_ops.h` and `model` class. For the `model` class, we can forward declare `friend` class, but there is no easy workaround for `raw_ops.h`.

Reply:
Hi @ljn917 

That's an interesting and great work. So, if I understood correctly, once `tf_tensor` is created with `get_tensor()` it gets linked with `tfe_handle` and there is no need to be updating it, right? That's very nice.

Regarding @ivanallen suggestion of setting `get_tensor()` private: right now is a bit difficult as all `raw_ops.h` depend on it. Perhaps we could use a `detail` namespace, declare `get_tensor()` on that namespace, and accessing it from `raw_ops.h`.

Reply:
@serizba Thanks for your comments.

> That's an interesting and great work. So, if I understood correctly, once `tf_tensor` is created with `get_tensor()` it gets linked with `tfe_handle` and there is no need to be updating it, right? That's very nice.

Yes, exactly. `tf_tensor` actually points to the internal data in `tfe_handle`.

> Regarding @ivanallen suggestion of setting `get_tensor()` private: right now is a bit difficult as all `raw_ops.h` depend on it. Perhaps we could use a `detail` namespace, declare `get_tensor()` on that namespace, and accessing it from `raw_ops.h`.

I am a bit confused by the `detail` namespace. `get_tensor()` is a public member function of `tensor` class. It acts as the getter method of the private `tf_tensor` member variable. We cannot simply move a member function to a namespace.

Reply:
@ljn917  

> Yes, exactly. tf_tensor actually points to the internal data in tfe_handle.

Then, the current implementation is perfect.

> I am a bit confused by the detail namespace. get_tensor() is a public member function of tensor class. It acts as the getter method of the private tf_tensor member variable. We cannot simply move a member function to a namespace.

Actually what I am proposing is a bit confusing and I didn't explain it properly. My suggestion is:

- Declaring `tf_tensor` private.
- Declaring `get_tensor()` outside the class definition, as an auxiliary function inside a `detail` namespace (before declaring `tensor`)
- Inside `tensor` declare `detail::get_tensor()` as a friend function.
- From outside, i.e. from `raw_ops.h`, call `detail::get_tensor()`.

As a schema, it would be as follows:

```c++

namespace cppflow {

    namespace detail {
        std::shared_ptr<TF_Tensor> get_tensor(tensor& t);
    }

    class tensor {
    public:
        ...
    private:
       
        // The original one
        std::shared_ptr<TF_Tensor> get_tensor();

        // The one that can be accessed from outside
        friend std::shared_ptr<TF_Tensor> detail::get_tensor(tensor& t);
     };
}

namespace cppflow {

    namespace detail {
        std::shared_ptr<TF_Tensor> detail::get_tensor(tensor& t) {
            // Actual definition
            return t.get_tensor();
        }
    }
    ...
}

```
Actually, the function `detail::get_tensor()` can still be accessed by users, but at least is somehow more protected, and users will know they are doing something nasty when accessing it.

I haven't tried this, but I think it may work. Nevertheless, I do not deny is somehow nasty, and not a very elegant solution. What do you think about it?

Reply:
Ah, I see what you mean. I think your solution works, but I really don't know which one is better...

Personally, I prefer to document the behaviors in comments and keep it as a member function. In fact, users are allowed to use it, e.g. for calling unimplemented tensorflow APIs. People just need to follow the contract and avoid undefined behaviors.

Reply:
Perfect,

I agree with you, I think it's fine to be able to access it, if it is clear what may happen. Indeed, as you pointed, it may be necessary for some users to access them.

Then, if the implementation is fine as it is right now, 
Should we merge it?

Reply:
Yes, I think it is ready to merge.